### DISCUSSION TITLE
Using 2p-kt programmatically

### DISCUSSION THREAD

**Discussion by Emmeral** [tuProlog/2p-kt#4] [Q&A]
I wanted to use the library programmatically from Java. Sadly I couldn't find any specific information regarding this in the `readme.md` or in the projects documentation (the [respective page]( https://pika-lab.gitlab.io/tuprolog/2p-in-kotlin/wiki/Users%20Guide/2p-programmatically-java/) is empty).

So my questions are:

1. Where is the programmatic entry point for adding a knowledge base and then querying the Prolog engine against this knowledge base. I have the knowledge base and the respective queries as a string representation.
2. Which part of the library do I need for that (core, parser, theory, ...?)

**Comment by gciatto** [comment]
Hi @Emmeral,

you are right, the doc is quite out of sync.

The most updated resource is [my beamer presentation](https://github.com/gciatto-unibo/2p-kt-presentation) (look into the release section to download the pdf).
It does not spend too many slides about Java, but there's something.
Are you comfortable with Kotlin-Java interoperability?

I'll answer to your questions below.

> Where is the programmatic entry point for adding a knowledge base and then querying the Prolog engine against this knowledge base. I have the knowledge base and the respective queries as a string representation

You need (at least) a `Solver` and a `ClauseParser` or `ClauseReader`.
The parser/reader is needed in order to parse your KB and queries.
The solver is needed to let Prolog answer queries.

> Which part of the library do I need for that (core, parser, theory, ...?)

You need modules `solve-classic` and `parser-theory` (and all their dependencis, thus `core`, `unify`, `theory`, and `parser-core`, which should be automatically downloaded if you use Gradle, or Maven).
Recall to add the `-jvm` suffix to artifact names when declaring dependencies on Gradle/Maven.

Hope this help & let me know if you need more info!
